llvm/tools/sycl-post-link/SpecConstants.cpp

```diff
@@ -521,16 +523,67 @@ PreservedAnalyses SpecConstantsPass::run(Module &M,
         //          %1, i32 %2), !SYCL_SPEC_CONST_SYM_ID !23
         // !23 = {!"string-id-2", i32 3, i32 4}
       } else {
-        // 2a. Spec constant must be resolved at compile time - just replace
-        // the intrinsic with default C++ value for the spec constant type.
-        Value *Default = getDefaultCPPValue(SCTy);
+        // 2a. Spec constant must be resolved at compile time - replace the
+        // intrinsic with the actual value for spec constant.
+        Value *Val = nullptr;
+        bool Is2020Intrinsic =
+            F.getName().startswith(SYCL_GET_SCALAR_2020_SPEC_CONST_VAL) ||
+            F.getName().startswith(SYCL_GET_COMPOSITE_2020_SPEC_CONST_VAL);
+
+        if (Is2020Intrinsic) {
+          // Handle SYCL2020 version of intrinsic - replace it with a pointer to
+          // specialization constant value.
+          // A pointer to a single RT-buffer with all the values of
+          // specialization constants is passed as a 3rd argument of intrinsic.
+          Value *RTBuffer =
+              IsComposite ? CI->getArgOperand(3) : CI->getArgOperand(2);
+
+          // Add the string literal to a "spec const string literal ID" ->
+          // "offset" map, uniquing the integer offsets if this is new
+          // literal.
+          auto Ins = OffsetMap.insert(std::make_pair(SymID, NextOffset));
+          bool IsNewSpecConstant = Ins.second;
+          auto CurrentOffset = Ins.first->second;
+          if (IsNewSpecConstant) {
+            if (IsComposite) {
+              // When handling elements of a structure, we do not use manually
+              // calculated offsets (which are sum of sizes of all previously
+              // encountered elements), but instead rely on data provided for us
+              // by DataLayout, because the structure can be unpacked, i.e.
+              // padded in order to ensure particular alignment of its elements.
+              auto *StructTy = cast<StructType>(
+                  CI->getArgOperand(0)->getType()->getPointerElementType());
+              const StructLayout *SL =
+                  M.getDataLayout().getStructLayout(StructTy);
+              NextOffset += SL->getSizeInBytes();
```

Comment:
We rely on host to have the same StructLayout for this spec constant RT values. Seems worth commenting this - here and then in the complimentary RT code which writes the values.
It would be great if we could somehow assert coincidence of the layouts.

BTW, it seems that we either need to limit 2020 spec constants to device copyable types only or invent another mechanism (field by field?) for value propagation from RT to device. + @AlexeySachkov

Reply:
> We rely on host to have the same StructLayout for this spec constant RT values.

That is correct, but this is also enforced by [SYCL specification](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:kernel.parameter.passing):

> The device compiler(s) must validate that the layout of any data shared between the host and the device(s) (e.g. value kernel arguments or data accessed through an accessor or USM) is compatible with the layout of that data on the host. If there is a layout mismatch that the implementation cannot or will not correct for (to make the layouts compatible), then the device compiler must issue an error and compilation must fail.
>
> 4.12.4. Rules for parameter passing to kernels

I assume that this task is already solved somewhere outside of `sycl-post-link` tool and here we can just rely on the fact that layouts are compatible.

> BTW, it seems that we either need to limit 2020 spec constants to device copyable types only or invent another mechanism (field by field?) for value propagation from RT to device. + @AlexeySachkov

This is a good point, but I guess even field-by-field copy won't help if user has defined non-trivial copy constructor for specialization constant. Tagging @gmlueck here for awareness

Reply:
> BTW, it seems that we either need to limit 2020 spec constants to device copyable types only or invent another mechanism (field by field?) for value propagation from RT to device.

Yes, good point.  I think the type needs to be "device copyable".  I've added this to my list of changes to propose when updating the section on `specialization_id`.